glib/src/subclass/shared.rs

```diff
@@ -0,0 +1,227 @@
+    /// The inner refcounted type
+    type RefCountedType: RefCounted;
+
+    /// Returns the type ID.
+    ///
+    /// This is usually defined via the [`Shared!`] derive macro.
+    ///
+    /// [`Shared!`]: ../../derive.Shared.html
+    fn get_type() -> crate::Type;
+
+    /// Converts the SharedType into its inner RefCountedType
+    fn into_refcounted(self) -> Self::RefCountedType;
+
+    /// Constructs a SharedType from a RefCountedType
+    fn from_refcounted(this: Self::RefCountedType) -> Self;
+}
+
+/// Register a boxed `glib::Type` ID for `T`.
+///
+/// This must be called only once and will panic on a second call.
+///
+/// See [`Shared!`] for defining a function that ensures that
+/// this is only called once and returns the type id.
+///
+/// [`Shared!`]: ../../derive.Shared.html
+pub fn register_shared_type<T: SharedType>() -> crate::Type {
+    unsafe {
+        use std::ffi::CString;
+        unsafe extern "C" fn shared_ref<T: SharedType>(v: ffi::gpointer) -> ffi::gpointer {
+            T::RefCountedType::ref_(v as *const <T::RefCountedType as RefCounted>::InnerType)
+                as ffi::gpointer
+        }
+        unsafe extern "C" fn shared_unref<T: SharedType>(v: ffi::gpointer) {
+            let _ = T::RefCountedType::from_ptr(
+                v as *const <T::RefCountedType as RefCounted>::InnerType,
+            );
+        }
+
+        let type_name = CString::new(T::NAME).unwrap();
+        if gobject_ffi::g_type_from_name(type_name.as_ptr()) != gobject_ffi::G_TYPE_INVALID {
+            panic!(
+                "Type {} has already been registered",
+                type_name.to_str().unwrap()
+            );
+        }
+
+        from_glib(gobject_ffi::g_boxed_type_register_static(
+            type_name.as_ptr(),
+            Some(shared_ref::<T>),
+            Some(shared_unref::<T>),
+        ))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    // Shared macro assumes 'glib' is in scope
+    use crate as glib;
+    use crate::value::ToValue;
+
+    #[derive(Clone, Debug, PartialEq, Eq)]
+    struct MySharedInner {
+        foo: String,
+    }
+
+    #[derive(Clone, Debug, PartialEq, Eq, glib::Shared)]
+    #[shared(type_name = "MySharedArc")]
+    struct MySharedArc(std::sync::Arc::<MySharedInner>);
+
+    #[derive(Clone, Debug, PartialEq, Eq, glib::Shared)]
+    #[shared(type_name = "MySharedRc")]
+    struct MySharedRc(std::rc::Rc::<MySharedInner>);
+
+    #[test]
+    fn test_register() {
+        assert_ne!(crate::Type::INVALID, MySharedArc::get_type());
+        assert_ne!(crate::Type::INVALID, MySharedRc::get_type());
+    }
+
+    #[test]
+    fn test_value_arc() {
+        assert_ne!(crate::Type::INVALID, MySharedArc::get_type());
+
+        let b = MySharedArc::from_refcounted(std::sync::Arc::new(MySharedInner {
+            foo: String::from("abc"),
+        }));
+        let v = b.to_value();
+        let b2 = v.get_some::<MySharedArc>().unwrap();
+        assert_eq!(b.0.foo, b2.0.foo);
```

Comment:
You probably also want to check everywhere for `Arc::ptr_eq` to make sure it's indeed the very same value and not just an equivalent copy